File: RenetikObjc/Library/Classes/Core/Categories/UIKit/UIBarButtonItem+CSExtension.m
//
// Created by Rene Dohan on 13/01/15.
// Copyright (c) 2015 creative_studio. All rights reserved.
//


#import "UIBarButtonItem+CSExtension.h"

@implementation UIBarButtonItem (CSExtension)

+ (UIBarButtonItem *)createWithImage:(UIImage *)image onClick:(void (^)(id sender))action {
    return [UIBarButtonItem.alloc bk_initWithImage:image style:UIBarButtonItemStylePlain handler:action];
}

+ (UIBarButtonItem *)createSoloTitle:(NSString *)title {
    return [UIBarButtonItem.alloc initWithTitle:title style:UIBarButtonItemStylePlain target:nil action:nil];
}

+ (UIBarButtonItem *)createWithTitle:(NSString *)title :(UIBarButtonItemStyle)style :(id)target :(SEL)action {
    return [UIBarButtonItem.alloc initWithTitle:title style:style target:target action:action];
}

+ (UIBarButtonItem *)createWithItem:(UIBarButtonSystemItem)item :(id)target :(SEL)action {
    return [UIBarButtonItem.alloc initWithBarButtonSystemItem:item target:target action:action];
}

+ (UIBarButtonItem *)createWithItem:(UIBarButtonSystemItem)item {
    return [UIBarButtonItem.alloc initWithBarButtonSystemItem:item target:nil action:nil];
}

+ (UIBarButtonItem *)createFlexSpaceItem {
    return [UIBarButtonItem.alloc initWithBarButtonSystemItem:UIBarButtonSystemItemFlexibleSpace target:nil action:nil];
}

- (void)setTarget:(id)target forAction:(SEL)action {
    self.target = target;
    self.action = action;
}

+ (UIImage *)imageFromSystemBarButton:(UIBarButtonSystemItem)systemItem :(UIColor *)color {
    UIToolbar *bar = UIToolbar.new;
    UIBarButtonItem *buttonItem = [UIBarButtonItem createWithItem:systemItem];
    [bar setItems:@[buttonItem] animated:NO];
    [bar snapshotViewAfterScreenUpdates:YES];
    for (UIView *view in [(id) buttonItem view].subviews)
        if ([view isKindOfClass:UIButton.class]) {
            UIImage *image = [((UIButton *) view).imageView.image imageWithRenderingMode:UIImageRenderingModeAlwaysTemplate];
            UIGraphicsBeginImageContextWithOptions(image.size, NO, image.scale);
            [color set];
            [image drawInRect:CGRectMake(0, 0, image.size.width, image.size.height)];
            image = UIGraphicsGetImageFromCurrentImageContext();
            UIGraphicsEndImageContext();
            return image;
        }
    return nil;
}


@end

File: Renetik/Renetik/Classes/Core/Extensions/UIKit/UIBarButtonItem+CSExtension.swift
//
// Created by Rene Dohan on 12/18/19.
//

import BlocksKit
import Renetik

public extension UIBarButtonItem {

    class var flexSpaceItem: UIBarButtonItem {
        UIBarButtonItem(barButtonSystemItem: .flexibleSpace, target: nil, action: nil)
    }

    class var fixedSpaceItem: UIBarButtonItem {
        UIBarButtonItem(barButtonSystemItem: .fixedSpace, target: nil, action: nil)
    }

    class func space(_ width: CGFloat = 15) -> UIBarButtonItem { fixedSpaceItem.width(width) }

    convenience init(image: UIImage, onClick: ArgFunc<UIBarButtonItem>? = nil) {
        self.init()
        if let action = onClick {
            bk_init(with: image, style: UIBarButtonItem.Style.plain, handler: { _ in action(self) })
        } else {
            self.init(image: image, style: UIBarButtonItem.Style.plain, target: nil, action: nil)
        }
    }

    convenience init(image: UIImage, onClick: Func? = nil) {
        self.init(image: image, onClick: { _ in onClick?() })
    }

    convenience init(view: UIView, onClick: ArgFunc<UIBarButtonItem>? = nil) {
        self.init(customView: view)
        if let action = onClick { view.onClick { action(self) } }
    }

    convenience init(view: UIView, onClick: @escaping Func) {
        self.init(customView: view)
        view.onClick { onClick() }
    }

    convenience init(item: UIBarButtonItem.SystemItem, onClick: ((_ sender: UIBarButtonItem) -> Void)? = nil) {
        self.init()
        if let action = onClick {
            bk_init(with: item, handler: { action($0 as! UIBarButtonItem) })
        } else {
            self.init(barButtonSystemItem: item, target: nil, action: nil)
        }
    }

    convenience init(title: String, onClick: ((_ sender: UIBarButtonItem) -> Void)? = nil) {
        self.init()
        if let action = onClick {
            bk_init(withTitle: title, style: UIBarButtonItem.Style.plain,
                    handler: { action($0 as! UIBarButtonItem) })
        } else {
            self.init(title: title, style: UIBarButtonItem.Style.plain, target: nil, action: nil)
        }
    }

    func width(_ value: CGFloat) -> Self { invoke { width = value } }
}